### PR TITLE
ZEPPELIN-501 Notebook search

### PR DESCRIPTION
**Pull request by bzz**
### What is this PR for?
This PR has for goal to allow the user to search through the code in all the paragraphs *and notebook names* in all the notebooks

It add a simple 'search bar' to the nav-bar of Zeppelin WebApp, and an in-memory fulltext search index of paragraphs to the backend.

The search is pretty basic now, fine-tuning it for better search over all types of source code will be a subject of further work.

### What type of PR is it?
Feature

### Todos
* [x] - Fix typos :dancer: b853aa6
* [x] - fix js issue in js console @felizbear 29da337
* [x] - update index on paragraph CRUD: 
  - [x] Read (initial work)
  - [x] Create\Delete 825b266
  - [x] Update e915a69
  - [x] Delete paragraph
* [x] - add license to zeppelin-distribution/src/bin_license/LICENSE (backend for lucene, not sure about angular-resource as it is part of the AngularJS codebase, but will add just in case) c00b516
* [x] - add missing Apache headers ded9c3b @felizbear 29da337
* [x] - fix CI (~~was failing RAT on zengine Too many files with unapproved license: 2~~, now flacky integration test AKA [ZEPPELIN-510](https://issues.apache.org/jira/browse/ZEPPELIN-510))
* [x] - index notebook names e80c3e5, @felizbear 29da337
* [x] - make NotebookRepoSync.sync() package private again  5a18bc8
* [x] - NPE on persisting changes of existing notebook
* [x] - reduce log verbosity 

### Is there a relevant Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-501

### How should this be tested?
Outline the steps to test the PR here.

### Screenshots (if appropriate)
screen shot 2015-12-15 at 19 49 18

screen shot 2015-12-15 at 19 52 03

### Questions:
* Does the licenses files need update? Yes, added to license files
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

This work is a collaboration with @felizbear who contributed major parts of the frontend changes.

### CODE REVIEW

**Comment by vgmartinez** [comment]
LGTM...
It is a good feature

**Comment by corneadoug** [comment]
@bzz Could you rewrite the PR description following the PR template (https://github.com/apache/incubator-zeppelin/blob/master/CONTRIBUTING.md#creating-a-pull-request)?

**Comment by bzz** [comment]
@vgmartinez  than you for positive feedback!

@corneadoug well, I expected a bit more of the code review, rather than "Wrong formatting of the PR description" ;)

I think I did a good job with JIRA issue. If you never the less feel that something is wrong - please contribute a new description you like better in the form of comment here and I will be willing to updated it.

**Comment by corneadoug** [comment]
@bzz Of course, the code review will come :)

Well, JIRA issue is JIRA issue, PR Template is PR Template :)
You will find that there is a lot of nice categories and questions that will help review the PR.

One good example is:
>Does the licenses files need update?

Which answer in this PR's case would be: Yes

But ``angular-resource`` was not added to the license file

File: docs/rest-api/rest-notebook.md [hunk 1]
```diff
@@ -424,4 +424,33 @@ limitations under the License.
       <td><pre>{"status":"OK","body":"* * * * * ?"}</pre></td>
     </tr>
   </table>
+
+  <table class="table-configuration">
+    <col width="200">
+    <tr>
+      <th>Full-text search thoug the paragraps in all notebooks</th>
```
Comment (AhyoungRyu) [inline docs/rest-api/rest-notebook.md]: @bzz I think there are some typos in this sentence! : )

File: zeppelin-server/pom.xml [hunk 2]
```diff
@@ -197,6 +197,22 @@
...
+
+    <dependency>
+      <groupId>javax.ws.rs</groupId>
+      <artifactId>javax.ws.rs-api</artifactId>
+      <version>2.0-m10</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.scala-lang</groupId>
+      <artifactId>scala-library</artifactId>
```
Comment (felixcheung) [inline zeppelin-server/pom.xml]: shouldn't this specify a version for the artifact?

**Comment by Leemoonsoo** [comment]
@bzz @felizbear This is cool feature!

I tried this branch and couple of feedbacks.

1. Error on notebook.controller.js after loading a notebook
image
2. Updating / Removing / Creating new notebook and paragraph does not update the index
3. I think indexing notebook title and paragraph title really make sense.
4. `zeppelin-distribution/src/bin_license/LICENSE` need to be updated

**Comment by bzz** [comment]
Thank you all guys for timely reviews and kind feedback!

TODOs:
 - [ ] tix fypos :dancer: 
 - [ ] fix js issue in js console
 - [ ] update index on paragraph CRUD
 - [ ] add license to `zeppelin-distribution/src/bin_license/LICENSE` (both, backend for lucene and front `angular-resource`)
 - [ ] add missing Apache headers
 - [ ] fix CI (now failing RAT on zengine `Too many files with unapproved license: 2` )

File: zeppelin-server/pom.xml (hunk 2 above)
Comment (bzz) [inline zeppelin-server/pom.xml]: Good catch, need to check if we do not specify it already somewhere.
Although it looks like a bit out of scope of this PR.

File: docs/rest-api/rest-notebook.md (hunk 1 above)
Comment (bzz) [inline docs/rest-api/rest-notebook.md]: Thank you for proof-reading! Added to the scope below

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java [hunk 3]
```diff
@@ -137,17 +137,14 @@ public void remove(String noteId) throws IOException {
    * copy new/updated notes from source to destination storage
    * @throws IOException
    */
-  void sync(int sourceRepoIndex, int destRepoIndex) throws IOException {
+  public void sync(int sourceRepoIndex, int destRepoIndex) throws IOException {
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: I think we don't need to expose this function as public api yet, the reason for declaring it as a `private-package` was to call it only from `sync()`  as well as for tests. Changing it to public may become reasonable when we start to support more than 2 backend repos simultaneously (surely logic should be changed as well).

Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: Good point, thank you for notice! It was changed through the implementation, but is not required any more, so will revert. Reflected in the scope above.

**Comment by khalidhuseynov** [comment]
thanks for a nice feature! one more question: in case of large number or notebooks (volume to be indexed), can there be any problem (e.g. system resources exhaustion) because of in-memory indexing (usage of `RAMDirectory`)? also any other performance issues possible

**Comment by corneadoug** [comment]
Here is my contribution for your PR Description

```
### What is this PR for?
This PR has for goal to allow the user to search through the code in all the paragraphs of all the notebooks
It add a simple 'search bar' to the nav-bar of Zeppelin WebApp, and an in-memory fulltext search index of paragraphs to the backend.

### What type of PR is it?
Feature

### Todos
* [x] - Fix fypos :dancer: b853aa6
* [ ] - fix js issue in js console
* [ ] - update index on paragraph CRUD
* [x] - add license to zeppelin-distribution/src/bin_license/LICENSE (backend for lucene, not sure about angular-resource as it is part of the AngularJS codebase, but will add just in case)
* [ ] - add missing Apache headers
* [ ] - fix CI (now failing RAT on zengine Too many files with unapproved license: 2 )
* [x] - index notebook names e80c3e5
* [x] - make NotebookRepoSync.sync() package private again

### Is there a relevant Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-501

### How should this be tested?
Outline the steps to test the PR here.

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? Yes, added to license files
* Is there breaking changes for older versions? No
* Does this needs documentation?
```

**Comment by bzz** [comment]
@khalidhuseynov 
>in case of large number or notebooks (volume to be indexed), can there be any problem (e.g. system resources exhaustion) because of in-memory indexing (usage of RAMDirectory)? also any other performance issues possible

Good question! As stated in PR description and JIRA issue:
>The search is pretty basic now, fine-tuning it for better search over all types of source code will be a subject of further work.

Performance, on-disk index, and code-search instead of word one are definitely those direction I'd love to see explored in further work.

**Comment by bzz** [comment]
@corneadoug thanks for description update, have moved everything there

**Comment by bzz** [comment]
CI fails now with unapproved licence, though last lines in log (`cat rat.txt`) do not have any reference of those...

```
[INFO] Rat check: Summary of files. Unapproved: 2 unknown: 2 generated: 0 approved: 6 licence.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Zeppelin ........................................... SUCCESS [  8.718 s]
[INFO] Zeppelin: Interpreter .............................. SUCCESS [01:17 min]
[INFO] Zeppelin: Zengine .................................. FAILURE [ 12.028 s]
```

Can anybody please verify that `mvn verify -Pusing-packaged-distr -Pspark-1.5 -Phadoop-2.3 -Ppyspark -B` passes?

As it looks like it hangs forever for me..

**Comment by bzz** [comment]
CI now fails with exactly the same case as on master under [ZEPPELIN-510](https://issues.apache.org/jira/browse/ZEPPELIN-510)

```
pabilities [{platform=LINUX, acceptSslCerts=true, javascriptEnabled=true, cssSelectorsEnabled=true, databaseEnabled=true, browserName=firefox, handlesAlerts=true, nativeEvents=false, webStorageEnabled=true, rotatable=false, locationContextEnabled=true, applicationCacheEnabled=true, takesScreenshot=true, version=31.0}]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:647)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:326)
	at org.openqa.selenium.remote.RemoteWebElement.sendKeys(RemoteWebElement.java:121)
	at org.apache.zeppelin.ZeppelinIT.testAngularDisplay(ZeppelinIT.java:240)
Caused by: org.openqa.selenium.ElementNotVisibleException: Element is not currently visible and so may not be interacted with
Build info: version: '2.48.2', revision: '41bccdd10cf2c0560f637404c2d96164b67d9d67', time: '2015-10-09 13:08:06'
System info: host: 'testing-worker-linux-docker-484929b1-3177-linux-6', ip: '172.17.4.152', os.name: 'Linux', os.arch: 'amd64', os.version: '3.13.0-40-generic', java.version: '1.7.0_76'
Driver info: driver.version: unknown
	at <anonymous class>.fxdriver.preconditions.visible(file:///tmp/anonymous6937570328313435477webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:9981)
	at <anonymous class>.DelayedCommand.prototype.checkPreconditions_(file:///tmp/anonymous6937570328313435477webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12517)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_/h(file:///tmp/anonymous6937570328313435477webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12534)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_(file:///tmp/anonymous6937570328313435477webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12539)
	at <anonymous class>.DelayedCommand.prototype.execute/<(file:///tmp/anonymous6937570328313435477webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12481)
```

**Comment by bzz** [comment]
History re-written to exclude accidental commit

**Comment by bzz** [comment]
@Leemoonsoo All reviews are addressed, I think it's ready to be merged.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/search/SearchService.java [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+          if (title != null) {
+            LOG.info("   Title: {}", doc.get("title"));
+          }
+
+          String text = doc.get(SEARCH_FIELD);
+          TokenStream tokenStream = TokenSources.getTokenStream(searcher.getIndexReader(), id,
+              SEARCH_FIELD, analyzer);
+          TextFragment[] frag = highlighter.getBestTextFragments(tokenStream, text, true, 3);
+          // TODO(bzz): remove this as too verbose
+          LOG.info("    {} fragments found for query '{}' in '{}'", frag.length, query, text);
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/search/SearchService.java]: We might not want to print contents of notebook in a log file. Either use LOG.debug or remove 'text' from the log.

**Comment by Leemoonsoo** [comment]
@bzz Great work.
Seems like index is not updated after paragraph remove. Is it possible to be handled?

**Comment by Leemoonsoo** [comment]
Following error is thrown when persisting changes of notebook which loaded on Zeppelin startup (not created after start up)

```
java.lang.NullPointerException
	at org.apache.zeppelin.notebook.Note.persist(Note.java:380)
	at org.apache.zeppelin.socket.NotebookServer.updateParagraph(NotebookServer.java:451)
	at org.apache.zeppelin.socket.NotebookServer.onMessage(NotebookServer.java:123)
	at org.apache.zeppelin.socket.NotebookSocket.onMessage(NotebookSocket.java:56)
	at org.eclipse.jetty.websocket.WebSocketConnectionRFC6455$WSFrameHandler.onFrame(WebSocketConnectionRFC6455.java:835)
	at org.eclipse.jetty.websocket.WebSocketParserRFC6455.parseNext(WebSocketParserRFC6455.java:349)
	at org.eclipse.jetty.websocket.WebSocketConnectionRFC6455.handle(WebSocketConnectionRFC6455.java:225)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:667)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by bzz** [comment]
@Leemoonsoo thank you for thought testing and review! Scope is updated, will address all the feedback, and let you know.

**Comment by bzz** [comment]
@Leemoonsoo all the feedback has been addressed now

**Comment by Leemoonsoo** [comment]
Tested and works really well. Thanks for really useful new feature.
LGTM!

**Comment by bzz** [comment]
Thank you! Merging if there are no other discussions
